### PR TITLE
PE-635: Safe TransactionID type

### PR DESCRIPTION
**Pull request by arielmelendez**

### CODE REVIEW

File: src/types/transaction_id.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+const trxIdRegex = /^([a-zA-Z]|[0-9]|-|_){43}$/;
```
Comment (matibat) [inline src/types/transaction_id.ts:1]: It can be simplified
```suggestion
const trxIdRegex = /^([a-zA-Z0-9]|-|_){43}$/;
```

Comment (matibat) [reply src/types/transaction_id.ts:1]: Also, confirm this: are underscores allowed?

Comment (matibat) [reply src/types/transaction_id.ts:1]: `\w` is the same than `[a-zA-Z0-9]|_`
```suggestion
const trxIdRegex = /^(\w|-){43}$/;
```

File: src/types/transaction_id.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+export class TransactionID {
+	constructor(private readonly transactionId: string) {
+		if (!transactionId.match(trxIdRegex)) {
+			throw new Error(
+				'Transaction ID should be a 43-character, alphanumeric string potentially including "=" and "_" characters.'
```
Comment (matibat) [inline src/types/transaction_id.ts:7]: > potentially including "="

Because it's a base64. Where do we trim the trailing equals?

Comment (matibat) [reply src/types/transaction_id.ts:7]: Should we accept it in the regex?

**Review by matibat — COMMENTED** [review]
Review in progress.
Left a comment for simplification and a few questions

File: src/arfs_entity_result_factory.ts [hunk 3]
```diff
@@ -1,5 +1,6 @@
 import { ArFSFileMetadataTransactionData } from './arfs_trx_data_types';
-import { TransactionID, DriveID, FolderID, FileID, FileKey, DriveKey } from './types';
+import { DriveID, FolderID, FileID, FileKey, DriveKey } from './types';
+import { TransactionID } from './types/transaction_id';
```
Comment (matibat) [inline src/arfs_entity_result_factory.ts:3]: Can't we export the safe types from `./types/index.ts`?

**Review by matibat — COMMENTED** [review]
I'm so happy to see these new types! :rocket:

File: src/types/transaction_id.ts (hunk 2 above)
Comment (arielmelendez) [reply src/types/transaction_id.ts:7]: I am following up on what number domain trx IDs are in.

File: src/types/transaction_id.ts (hunk 1 above)
Comment (arielmelendez) [reply src/types/transaction_id.ts]: 46f46fe

File: src/types/transaction_id.ts (hunk 2 above)
Comment (arielmelendez) [reply src/types/transaction_id.ts:9]: Based on convos with Arweave team in Discord it seems that =, though a valid Base64 URL character, would only appear at the end of a hashed string, and only if padding were required. Apparently no padding is required for the hashes we create here.

File: src/arfs_entity_result_factory.ts (hunk 3 above)
Comment (arielmelendez) [reply src/arfs_entity_result_factory.ts]: 3824924
Exported as a folder.

File: src/types/transaction_id.ts (hunk 2 above)
Comment (matibat) [reply src/types/transaction_id.ts:9]: We are covered here. The network doesn't require the trailing equals

File: src/arfs_entity_result_factory.ts (hunk 3 above)
Comment (matibat) [reply src/arfs_entity_result_factory.ts]: It looks better! **-** We are having conflicts importing between `src/types.ts` and `src/types/`. Looking forward to that change

Comment (matibat) [reply src/arfs_entity_result_factory.ts]: Maybe I'd expect that change at #130 or further PRs for PE-635

**Review by matibat — APPROVED** [review]
It's looking good to me!
Just looking forward to the upcoming `types` import fix

File: src/arfs_entity_result_factory.ts (hunk 3 above)
Comment (arielmelendez) [reply src/arfs_entity_result_factory.ts]: Yea I'll take care of it for ByteCount

**Review by fedellen — APPROVED** [review]
👍

**Pull request merged**